core/src/main/java/com/graphhopper/routing/util/EncodingManager.java

```diff
@@ -20,99 +20,311 @@
+    private int bitsForEdgeFlags;
     private final int bitsForTurnFlags = 8 * 4;
-    private int nextWayBit = 0;
-    private int nextNodeBit = 0;
-    private int nextRelBit = 0;
-    private int nextTurnBit = 0;
     private boolean enableInstructions = true;
     private String preferredLanguage = "";
 
     /**
-     * Instantiate manager with the given list of encoders. The manager knows several default
-     * encoders ignoring case.
-     * <p>
+     * This constructor creates the object that orchestrates the edge properties.
      *
-     * @param flagEncodersStr comma delimited list of encoders. The order does not matter.
+     * @param extendedDataSpaceInBytes in bytes
      */
-    public EncodingManager(String flagEncodersStr) {
-        this(flagEncodersStr, 4);
+    private EncodingManager(TagsParser parser, int extendedDataSpaceInBytes) {
+        this.parser = parser;
+        this.extendedDataSize = Math.max(1, extendedDataSpaceInBytes / 4) * 4;
     }
 
-    public EncodingManager(String flagEncodersStr, int bytesForEdgeFlags) {
-        this(FlagEncoderFactory.DEFAULT, flagEncodersStr, bytesForEdgeFlags);
+    public IntsRef createIntsRef() {
+        return new IntsRef(extendedDataSize / 4);
+    }
+
+    public static class Builder {
+        private boolean buildCalled;
+        private final EncodingManager em;
+        private int nextNodeBit = 0;
+        private int nextRelBit = 0;
+        private int nextTurnBit = 0;
+
+        /**
+         * used only for tests, could be removed in 0.11
+         */
+        public Builder() {
+            this(new TagsParser(), 8);
+        }
+
+        public Builder(TagsParser parser, int extendedDataSize) {
+            this.em = new EncodingManager(parser, extendedDataSize);
+        }
+
+        /**
+         * This method adds some EncodedValues that are required like roundabout and road_class
+         */
+        public Builder addGlobalEncodedValues(boolean surface, boolean carMaxSpeed) {
+            // TODO NOW for all bike we need surface (unpaved) as global encoded value to avoid creating multiple -> addBikeEncodedValues()?
+            if (surface) {
+                List<String> surfaces = Arrays.asList("_default", "paved", "asphalt", "cobblestone", "cobblestone:flattened", "sett", "concrete",
+                        "concrete:lanes", "concrete:plates", "paving_stones", "paving_stones:30", "unpaved", "compacted"
+                        , "dirt", "earth", "fine_gravel", "grass", "grass_paver", "gravel", "ground", "ice", "metal"
+                        , "mud", "pebblestone", "salt", "sand", "wood");
+                add(TagParserFactory.createSurface(new StringEncodedValue(TagParserFactory.SURFACE, surfaces, "_default")));
+            }
+
+            if (carMaxSpeed) {
+                // TODO NOW do we really need this for instructions?
+                DecimalEncodedValue ev = new DecimalEncodedValue(TagParserFactory.CAR_MAX_SPEED, 7, 0, 2, false);
```

Comment:
Would this be a duplicate of
`add(TagParserFactory.Car.createMaxSpeed(new DecimalEncodedValue(TagParserFactory.CAR_MAX_SPEED, 5, 0, 5, false), filter)).`? I would assume that one is the maxspeed for cars and the other the tagged max speed? But why is it car max speed bot htimes? Or is this just not implemented yet :)?

Reply:
Ah, ok just read the next line :). `add(TagParserFactory.Car.createAverageSpeed(new DecimalEncodedValue(TagParserFactory.CAR_AVERAGE_SPEED, 5, 0, 5, false), speedMap)).`

I guess the idea woul dbe to check if both speed would be compatible and if yes, only store it once. In this case the number of bits and the factor would not match.

Reply:
Ah ok the `OSMReaderTest#testEncodedValueBasedEncodingManager` actually fails because of this. `Cannot add parser car.max_speed. Already existing: car.max_speed`

Reply:
After removing this line there is another exception: `java.lang.RuntimeException: Graph after reading OSM must not be empty. Read 12 items and 0 locations`

Reply:
Ok, the issue seems to be that the access TagParser is not yet used in the EncodingManager. Currently the EM only checks other FlagEncoders.

Reply:
Will have a look into OSMReaderTest to see what you mean

Reply:
For me the testEncodedValueBasedEncodingManager passes. What do you mean with "the access TagParser is not yet used in the EncodingManager"?

Reply:
Interesting, maybe IntelliJ did something unexpected here. When I checked out the branch and tried the test, it failed for me. I just updated the branch to use your latest changes and then it wouldn't even build anymore. Running `mvn clean` fixed the issue, probably something changed that IntelliJ couldn't cope with. The test works now for me as well.

> What do you mean with "the access TagParser is not yet used in the EncodingManager"?

When I debugged the test to see what was happening, the EncodingManager didn't use the access tag parser in `acceptWay`. Now I does, but I can see that you didn't add it afterwards. I am not exactly sure what happened here. Sorry for the troubles.